### PR TITLE
Add strace qemu process support

### PR DESCRIPTION
**Pull request by xutian**
Signed-off-by: Xu Tian <xutian@redhat.com>

### CODE REVIEW

File: virttest/test_setup.py [hunk 1]
```diff
@@ -1767,3 +1769,69 @@ def _all_killed():
...
+    It's useful to analyze qemu hang issue. But it will generate
+    a large size logfile if it is enabled, so compressed original
+    logfiles after strace process terminated to safe disk space.
+    """
+
+    def __init__(self, test, params, env):
+        self.env = env
+        self.test = test
+        self.params = params
+        self.tarball = "strace.tgz"
```
Comment (PandaWei) [inline virttest/test_setup.py]: Is the variable 'tarball' used as log tar, right ?
If so, It's better to rename it.

Comment (xutian) [inline virttest/test_setup.py]: yes.  it's a log tarball.   Renamed to make  it more readable. 

thanks,
Xu

File: virttest/test_setup.py [hunk 2]
```diff
@@ -1767,3 +1769,69 @@ def _all_killed():
...
+    a large size logfile if it is enabled, so compressed original
+    logfiles after strace process terminated to safe disk space.
+    """
+
+    def __init__(self, test, params, env):
+        self.env = env
+        self.test = test
+        self.params = params
+        self.process = path.findcommand("strace")
+        self.log_tarball = "strace.tgz"
```
Comment (suqinhuang) [inline virttest/test_setup.py]: no attribute 'findcommand'

**Comment by xutian** [comment]
oh, my fault. a typo.  fixed.  Hi @suqinhuang , would you like to review again.

File: virttest/test_setup.py [hunk 3]
```diff
@@ -1767,3 +1769,69 @@ def _all_killed():
...
+    def root_dir(self):
+        root_dir = os.path.join(self.test.debugdir, "strace")
+        if not os.path.isdir(root_dir):
+            os.makedirs(root_dir)
+        return root_dir
+
+    def _generate_cmd(self, vm):
+        """Generate strace start command line"""
+        pid = vm.get_pid()
+        template = "{strace} -p {pid} -f -F -D -T -C -W  -o {logfile}"
```
Comment (suqinhuang) [inline virttest/test_setup.py]: invalid option -- 'W'

File: virttest/test_setup.py [hunk 4]
```diff
@@ -1767,3 +1769,69 @@ def _all_killed():
...
+        kwargs = {"strace": self.process,
+                  "pid": pid,
+                  "logfile": logfile}
+        return template.format(**kwargs)
+
+    def _start(self, cmd):
+        """Start strace process in sub-process"""
+        sub_process = process.SubProcess(cmd, shell=True)
+        sub_process.start()
+        if self.env["strace_processes"]:
```
Comment (suqinhuang) [inline virttest/test_setup.py]: 11:13:02 stacktrace       L0042 ERROR|   File "/root/WorkFolder/avocado/avocado/virttest/test_setup.py", line 1811, in _start
11:13:02 stacktrace       L0042 ERROR|     if self.env["strace_processes"]:
11:13:02 stacktrace       L0042 ERROR|   File "/usr/lib64/python2.7/UserDict.py", line 23, in __getitem__
11:13:02 stacktrace       L0042 ERROR|     raise KeyError(key)
11:13:02 stacktrace       L0042 ERROR| KeyError: 'strace_processes'

**Comment by xutian** [comment]
[root@xutian-dev cfg]# /root/tests/avocado/scripts/avocado run --vt-config=tests-example.cfg 
JOB ID     : 0a38026eb4003f8e5b6e25b8dae9e5cc190f48d7
JOB LOG    : /root/avocado/job-results/job-2015-12-01T17.28-0a38026/job.log
TESTS      : 1
 (1/1) smp_2.4096m.repeat1.run_test.Host_RHEL.m6.7.qcow2.virtio_blk.up.virtio_net.RHEL.6.5.x86_64.io-github-autotest-qemu.boot: PASS (77.29 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2015-12-01T17.28-0a38026/html/results.html
TIME       : 77.29 s
[root@xutian-dev cfg]# ls -l  /root/avocado/job-results/job-2015-12-01T17.28-0a38026/test-results/smp_2.4096m.repeat1.run_test.Host_RHEL.m6.7.qcow2.virtio_blk.up.virtio_net.RHEL.6.5.x86_64.io-github-autotest-qemu.boot/
total 248
drwxr-xr-x 2 root root  4096 Dec  1 17:28 data
-rw-r--r-- 1 root root 51044 Dec  1 17:29 debug.log
-rw-r--r-- 1 root root 49401 Dec  1 17:29 monitor-catch_monitor-20151201-172820-l6QUKWmW.log
-rw-r--r-- 1 root root 15451 Dec  1 17:29 monitor-qmpmonitor1-20151201-172820-l6QUKWmW.log
-rw-r--r-- 1 root root     0 Dec  1 17:28 remote.log
drwxr-xr-x 2 root root  4096 Dec  1 17:29 screendumps_avocado-vt-vm1_2919_iter0
-rw-r--r-- 1 root root   300 Dec  1 17:29 screendumps_avocado-vt-vm1_2919_iter0.webm
-rw-r--r-- 1 root root  4305 Dec  1 17:28 seabios-avocado-vt-vm1.log
-rw-r--r-- 1 root root 34779 Dec  1 17:29 serial-serial0-avocado-vt-vm1.log
-rw-r--r-- 1 root root   316 Dec  1 17:29 session-avocado-vt-vm1-2J9U.log
-rw-r--r-- 1 root root   340 Dec  1 17:29 session-avocado-vt-vm1-x4mh.log
-rw-r--r-- 1 root root  5389 Dec  1 17:29 stderr
-rw-r--r-- 1 root root    26 Dec  1 17:29 stdout
-rw-r--r-- 1 root root   312 Dec  1 17:29 strace.tgz
drwxr-xr-x 5 root root  4096 Dec  1 17:28 sysinfo
-rw-r--r-- 1 root root 34273 Dec  1 17:29 tcpdump.log
drwxr-xr-x 2 root root  4096 Dec  1 17:29 vm_register_avocado-vt-vm1_2919
-rw-r--r-- 1 root root    36 Dec  1 17:29 whiteboard

**Comment by lmr** [comment]
@xutian can you please rebase against latest master? That will make the CI problem to go away, thanks!

**Comment by xutian** [comment]
@lmr  thanks your reminder, PR updated. 

Thanks,
Xu

**Comment by suqinhuang** [comment]
LGTM

Acked-by: Suqin Huang

**Comment by suqinhuang** [comment]
ID: 1285983

**Comment by xutian** [comment]
Merged.  Thanks @suqinhuang

**Pull request merged**